### DISCUSSION TITLE
Customizing the new account console v3 theme

### DISCUSSION THREAD

**Discussion by panda1846** [keycloak/keycloak#29044] [Q&A]
Hi, 

In a project I'm currently working on, I need to customize the account theme (v3) of the latest version of Keycloak and I've run into some issues. I have used the Keycloak quickstarts as a guide (https://github.com/keycloak/keycloak-quickstarts/tree/latest/extension/extend-account-console), but the example does not show how to solve my problems. 

First issue: I need to access custom environment variables in my added views. In previous versions of the account theme, you could add custom values in the theme.properties file (myValue=${env.MY_VALUE}) and access them in the .ftl templates (with properties.myValue). How can I achieve this with the new account theme? When I try to add a customized index.ftl in my .jar, the account console is not loading anymore because scripts are missing.

Second issue: I need to get a Keycloak token in my React component because I need to send requests to custom endpoints implemented in another Keycloak extension. How can I use the Keycloak Js library in my custom component for the account console? The set up with keycloak.init()... caused an infinite loop when rendering my component in the account console

Thanks in advance

**Comment by codespearhead** [comment]
[This answer](https://github.com/keycloak/keycloak/discussions/28393#discussioncomment-9010960) still holds true: you have to use Apache Freemarker for markup and custom SPIs for making data available in the markup, as shown below via variable `user`.

https://github.com/keycloak/keycloak/blob/e93b7d4f3a55c213ff0dc6f3ea299d4f2ed24eca/services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java#L187-L192

https://github.com/keycloak/keycloak/blob/e93b7d4f3a55c213ff0dc6f3ea299d4f2ed24eca/services/src/main/java/org/keycloak/forms/login/freemarker/Templates.java#L47-L48

https://github.com/keycloak/keycloak/blob/e93b7d4f3a55c213ff0dc6f3ea299d4f2ed24eca/themes/src/main/resources/theme/base/login/login-verify-email.ftl#L6

I'm not sure I understand the second question. `keycloak.init()` probably caused an infinite loop because it's [already available](https://github.com/keycloak/keycloak/blob/e93b7d4f3a55c213ff0dc6f3ea299d4f2ed24eca/js/apps/account-ui/src/root/KeycloakContext.tsx#L65) there.

Possible Duplicates (or related issues): #28393 , #29226 , #28282 , #22663 #29232 .

**Reply by patmoshosting** [reply]
The second question is this: how do I get an access token for the currently logged-in user from a React component in a custom theme? This seems like it should be possible, but there is no documentation anywhere on how to do it.
I tried to import the `useEnvironment` hook from the `KeycloakContext` file you linked to, but I can't figure out what package I'm supposed to import it from.

Edit: it looks like Keycloak's theme files use `import { useEnvironment } from "@keycloak/keycloak-ui-shared";`, but that's not working for me. If I try to call `useEnvironment` without adding a `KeycloakProvider` in my page (under the assumption that there is a parent component somewhere that has the `KeycloakProvider`), I get an error `Cannot read properties of undefined (reading '_currentValue')`. If I instead add a `KeycloakProvider` to my page, the page just refreshes itself in an infinite loop.
